preemptive_requests.py

```diff
@@ -35,23 +77,362 @@ def setup_requests(self):
         return 'preemptive_requests.ui'
+
+    def channels(self):
+        """
+        Make sure PyDM can find the channels we set up for cleanup.
+
+        Put this here instead of on the PMPSTableWidgetItem because
+        QTableWidgetItem instances are not instances of QWidget, and
+        therefore are not checked by PyDM for channels.
+        """
+        return self._channels
+
+
+class PMPSTableWidgetItem(QTableWidgetItem):
+    """
+    QTableWidgetItem with extra utilities for the PMPS UI
+
+    Adds the following features:
+    - Fill value and connection state from PyDMChannel
+    - Configurable sorting
+    - Process inputs before storing in the table
+    - Able to sort as a non-str type
+
+    Parameters
+    ----------
+    store_type : callable
+        Function or type to call on the value from the PyDMChannel before
+        storing in the table.
+    data_type : callable
+        Function or type to call on the str stored in the table before
+        comparing with other PMPSTableWidgetItem instances for sorting.
+    default : Any
+        A starting value for the widget item.
+    channel : str, optional
+        PyDM channel address for value and connection updates.
+    """
+    def __init__(self, store_type, data_type, default,
+                 channel=None, parent=None):
+        QTableWidgetItem.__init__(self, parent)
+        self.store_type = store_type
+        self.data_type = data_type
+        self.setText(str(default))
+        self.channel = channel
+        self.connected = False
+        if channel is not None:
+            self.pydm_channel = PyDMChannel(
+                channel,
+                value_slot=self.update_value,
+                connection_slot=self.update_connection,
+                )
+            self.pydm_channel.connect()
+
+    def update_value(self, value):
+        """
+        Use the hidden text field to store the value from the PV.
+
+        This is pre-processed by the "store_type" attribute because the
+        value can only be saved as a string via setText.
+
+        We use the setText instead of an attribute to help with debugging,
+        this means you can see what the table is being sorted on
+        if you unhide the columns.
+        """
+        self.setText(str(self.store_type(value)))
+
+    def update_connection(self, connected):
+        """
+        When our PV connects or disconnects, store the state as an attribute.
+        """
+        self.connected = connected
+
+    def get_value(self):
+        """The value in canonical python type (not string)."""
+        return self.data_type(self.text())
+
+    def __lt__(self, other):
+        """Use order of defined type, not alphabetical."""
+        return self.get_value() < other.get_value()
+
+
+def bitmask_count(bitmask):
+    """Count the number of high bits in a bitmask."""
+    bitmask = int(bitmask)
+    count = 0
+    if bitmask < 0:
+        bitmask += 2**32
+    while bitmask > 0:
```

Comment:
Lazy me sees this and thinks, hmm... `str(bin(bitmask)).count('1')`

Reply:
Ah! This is what happens when I code in non-python, I forget that there's a more straighforward way to do everything.